storage/pipeline/checks.go

```diff
@@ -51,43 +52,84 @@ func checkPieces(ctx context.Context, maddr address.Address, sn abi.SectorNumber
 	var offset abi.PaddedPieceSize
 
 	for i, p := range pieces {
+		p, i := p, i
+
 		// check that the piece is correctly aligned
-		if offset%p.Piece.Size != 0 {
-			return &ErrInvalidPiece{xerrors.Errorf("sector %d piece %d is not aligned: size=%xh offset=%xh off-by=%xh", sn, i, p.Piece.Size, offset, offset%p.Piece.Size)}
-		}
-		offset += p.Piece.Size
-
-		// if no deal is associated with the piece, ensure that we added it as
-		// filler (i.e. ensure that it has a zero PieceCID)
-		if p.DealInfo == nil {
-			exp := zerocomm.ZeroPieceCommitment(p.Piece.Size.Unpadded())
-			if !p.Piece.PieceCID.Equals(exp) {
-				return &ErrInvalidPiece{xerrors.Errorf("sector %d piece %d had non-zero PieceCID %+v", sn, i, p.Piece.PieceCID)}
-			}
-			continue
+		if offset%p.Piece().Size != 0 {
+			return &ErrInvalidPiece{xerrors.Errorf("sector %d piece %d is not aligned: size=%xh offset=%xh off-by=%xh", sn, i, p.Piece().Size, offset, offset%p.Piece().Size)}
 		}
-
-		dealCount++
-
-		deal, err := api.StateMarketStorageDeal(ctx, p.DealInfo.DealID, ts.Key())
+		offset += p.Piece().Size
+
+		err := p.handleDealInfo(handleDealInfoParams{
+			FillerHandler: func(pi UniversalPieceInfo) error {
+				// if no deal is associated with the piece, ensure that we added it as
+				// filler (i.e. ensure that it has a zero PieceCID)
+
+				exp := zerocomm.ZeroPieceCommitment(p.Piece().Size.Unpadded())
```

Comment:
@magik6k I'm curious, Is there a specific reason we’re not allowing non-zero piece here? It might be CC sector made out of random data instead of zeroes.That should be a valid scenario from the perspective of a protocol. 

Making assumptions that each CC sector is made of zero-ed input will be even more wrong after introducing sector re-snapping - from the perspective of lotus, each sector without active deals is a CC sector. 
Having non-zero commd will be a common scenario for CC in the future, deal sectors may become cc sectors just by having a longer lifetime than a deal, and be snaped-up from that state again

Reply:
As I understand, any non-zero piece is a ddo piece/sector (which doesn't require a PAM, just piece data, so from the point of view of the actor it looks like a sector with non-zero data).

This is because all-zero sectors have two additional properties:
* You can do snap deals on them (this doesn't really change with resnap probably, data will likely still be encoded into 'sector key' which is the same as all-zero sector)
* You don't have to specify unsealed CID in chain messages (zero commd is implicit/default), so onboarding empty CC sectors is slightly cheaper